### PR TITLE
Light material bit masking

### PR DESCRIPTION
**Pull request by guycalledfrank**
Light mask
A bitmask to select which lights affect which surfaces. New API:
- light.mask
- material.mask

Light affects material if (light.mask & material.mask) > 0

### CODE REVIEW

File: src/framework/components/light/light_system.js [hunk 1]
```diff
@@ -163,6 +163,12 @@ pc.extend(pc, function () {
                 type: ['point', 'spot']
             }
         }, {
+            name: "mask",
+            displayName: "Mask",
+            description: "Bit mask",
+            type: "number",
+            defaultValue: 1,
```
Comment (daredevildave) [inline src/framework/components/light/light_system.js]: expose: false

Comment (guycalledfrank) [inline src/framework/components/light/light_system.js]: Would be useful to have in UI though. Unity has UI for bitmask things.

**Comment by willeastcott** [comment]
Are bit masks user friendly? Don't I want to create a tag called 'Lightmapped' and set it on a material and exclude the sunlight directional from 'Lightmapped'.

**Comment by guycalledfrank** [comment]
You can give user names to bits in UI and it'll be user-friendly.

File: src/framework/components/light/light_data.js [hunk 2]
```diff
@@ -14,6 +14,7 @@ pc.extend(pc, function () {
         this.innerConeAngle = 40;
         this.outerConeAngle = 45;
         this.falloffMode = pc.LIGHTFALLOFF_LINEAR;
+        this.mask = 1;
```
Comment (willeastcott) [inline src/framework/components/light/light_data.js]: Shouldn't this be in the pc.Light structure?

File: src/scene/scene_phongmaterial.js [hunk 3]
```diff
@@ -210,6 +210,7 @@ pc.extend(pc, function () {
             this.specularMapTint = false;
             this.emissiveMapTint = false;
             this.emissiveIntensity = 1;
+            this.mask = 1;
```
Comment (willeastcott) [inline src/scene/scene_phongmaterial.js]: Should this be in pc.Material? Shouldn't all materials be able to specify which lights they are lit by? Not just Phong Materials?

File: src/framework/components/light/light_data.js (hunk 2 above)
Comment (guycalledfrank) [inline src/framework/components/light/light_data.js]: it is in all possible structures, isn't it?

File: src/scene/scene_phongmaterial.js (hunk 3 above)
Comment (guycalledfrank) [inline src/scene/scene_phongmaterial.js]: Maybe, but shader generation and light usage in shader depends on shader generator implementation.

Comment (willeastcott) [inline src/scene/scene_phongmaterial.js]: Yep. But right now, if you wrote another material based on a shader generator, you'd need this in the pc.Material base class. Seems like it belongs there to me.

File: src/framework/components/light/light_data.js (hunk 2 above)
Comment (willeastcott) [inline src/framework/components/light/light_data.js]: As I may have mentioned before, I would like pc.Light and pc.Camera to be merged into the Light and Camera components at some point soon. Putting 'mask' into the LightComponent is like you've started doing this already. The current convention is that all light specific properties go in pc.Light. So it's not wrong, but it doesn't conform to the current 'architecture'. ;)

Comment (willeastcott) [inline src/framework/components/light/light_data.js]: Oh, actually, I see you do save the mask property into the light. But mask is not initialized in the pc.Light constructor. I think that's my problem with it. :)

File: src/scene/scene_phongmaterial.js (hunk 3 above)
Comment (guycalledfrank) [inline src/scene/scene_phongmaterial.js]: fixed

File: src/framework/components/light/light_data.js (hunk 2 above)
Comment (guycalledfrank) [inline src/framework/components/light/light_data.js]: fixed, turns out I just forgot to add the file initially

**Pull request merged**